addons/uom/models/uom_uom.py

```diff
@@ -166,3 +211,40 @@ def _compute_price(self, price, to_unit):
         if to_unit:
             amount = amount / to_unit.factor
         return amount
+
+    def _filter_protected_uoms(self, unlink=False):
+        """Verifies self does not contain protected uoms."""
+        if self.env.context.get('install_mode'):
+            return self.browse()
+        linked_model_data = self.env['ir.model.data'].search([
+            ('model', '=', self._name),
+            ('res_id', 'in', self.ids),
+            ('name', 'in', PROTECTED_XML_IDS),
+        ])
+        if not linked_model_data:
+            return self.browse()
+        elif unlink:
+            # Main uoms are protected against deletion, even if not used in fields.
+            return self.browse(set(linked_model_data.mapped('res_id')))
+        uom_fields = self.env['ir.model.fields'].search([
+            ('ttype', '=', 'many2one'),
+            ('state', '=', 'base'),
+            ('related', '=', False),
+            ('store', '=', True),
+            ('compute', '=', False),
+            ('relation', '=', 'uom.uom'),
+        ])
+        used_uom_ids = set()
+        for field in uom_fields:
+            model_env = self.env[field.model]
+            if not model_env._auto:
+                # Do not consider reporting views
+                continue
+            query = """
+                SELECT DISTINCT {uom_field}
+                from {model}
+                where {uom_field} in %s
+            """.format(uom_field=field.name, model=model_env._table)
+            self.env.cr.execute(query, [tuple(self.ids)])
+            used_uom_ids |= {res[0] for res in self.env.cr.fetchall() if res[0]}
```

Comment:
Hmm that's a recipe for horrible performance. None of those fields are indexed and most of the FKs will be in the large tables. Also why don't we replace this with `ondelete=restrict` for every m2o that points to a uom - feels much simpler and cleaner.

Reply:
This is globally already the case since the majority of the uom fields are required and `ondelete="restrict"`.